### PR TITLE
DRA kubelet: separate beta and alpha gRPC APIs

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Reusing types from the alpha in the beta made it possible to provide and use both versions without conversion. The downside was that removal of the alpha would have been harder, if not impossible. DRA drivers could continue to use the alpha types and provided the beta interface automatically.

Now the two versions are completely separate gRPC APIs, although in practice there are no differences besides the name. Support for the alpha API in kubelet is provided via automatically generated conversion and manually written interface wrappers.

Those are provided as part of the v1alpha4 package. The advantage of having all of that in a central place is that it'll be easier to remove when no longer needed.

#### Does this PR introduce a user-facing change?
```release-note
DRA: DRA driver authors should update their DRA drivers to use the v1beta1 gRPC API. The older alpha API still works, but might get removed eventually.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
```

/assign @SergeyKanzhelev

### CODE REVIEW

**Review by SergeyKanzhelev — APPROVED** [review]
This is great! I love that all the conversion tax is on alpha API.

I have a removal code if needed, but what you did is actually the best.

/lgtm
/hold

please unhold at will

**Comment by pohly** [comment]
@klueska, @bart0sh, @byako : any comments?

If not, then I'll lift the hold later today to get this merged.

**Comment by klueska** [comment]
Yes. This is better
/approve
/lgtm

File: staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go [hunk 1]
```diff
@@ -272,12 +272,12 @@ type draPlugin struct {
...
-// with Kubernetes >= 1.32. To be compatible with Kubernetes >= 1.31, a driver
-// has to ask specifically to register only the alpha gRPC API, i.e. use:
-//
-//	Start(..., NodeV1beta1(false))
-func Start(ctx context.Context, nodeServer interface{}, opts ...Option) (result DRAPlugin, finalErr error) {
+// The DRA driver decides which gRPC interfaces it implements. At least one
+// implementation of [drapbv1alpha4.NodeServer] or [drapbv1beta1.DRAPluginServer]
+// is required. Implementing drapbv1beta1.DRAPluginServer is recommended for
+// DRA driver targeting Kubernetes >= 1.32. To be compatible with Kubernetes 1.31,
+// DRA drivers must implement only [drapbv1alpha4.NodeServer].
```
Comment (aojea) [inline staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:279]: Is this last sentence correct? it seems to contradict the previous if "DRA drivers must implement only" the alpha version

File: staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go [hunk 2]
```diff
@@ -338,15 +338,17 @@ func Start(ctx context.Context, nodeServer interface{}, opts ...Option) (result
...
-		if nodeServer, ok := nodeServer.(drapbv1alpha4.NodeServer); ok && o.nodeV1alpha4 {
-			logger.V(5).Info("registering v1alpha4.Node gGRPC service")
-			drapbv1alpha4.RegisterNodeServer(grpcServer, nodeServer)
-			supportedServices = append(supportedServices, drapbv1alpha4.NodeService)
-		}
-		if nodeServer, ok := nodeServer.(drapbv1beta1.DRAPluginServer); ok && o.nodeV1beta1 {
-			logger.V(5).Info("registering v1beta1.DRAPlugin gRPC service")
-			drapbv1beta1.RegisterDRAPluginServer(grpcServer, nodeServer)
-			supportedServices = append(supportedServices, drapbv1beta1.DRAPluginService)
+		for _, nodeServer := range nodeServers {
```
Comment (aojea) [inline staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:341]: do we want to provide priority or left to the developer to pass the order of preference of the version they prefer?

File: staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go (hunk 1 above)
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:279]: The last sentence *is* "DRA drivers must implement only". Can you clarify which two sentences contradict each other?

File: staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go (hunk 2 above)
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:341]: The code registers everything that is passed in and enabled via options. It doesn't matter which server gets registered first (alpha or beta).

Theoretically, a driver could pass an implementation for the same version twice. I haven't tested it, but I suppose registration then will barf or one server overwrite the other.

File: staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go (hunk 1 above)
Comment (aojea) [reply staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:279]: oh, my bad, missed the comma 😄

File: staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go (hunk 2 above)
Comment (aojea) [reply staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:341]: no worries, was just a curiosity question

File: staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go (hunk 1 above)
Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:279]: Agreed, that is a bit subtle. Perhaps starting that sentence on a new line would have been more obvious, but I'll merge it now as-is.

/hold cancel

**Pull request merged**